### PR TITLE
Fix installation object parsing logic

### PR DESCRIPTION
**Pull request by hanxiaop**
**Please provide a description of this PR:**
Related to https://github.com/istio/istio/pull/45654

The original parsing logic may not be very robust,
for example:
```
apiVersion: v1
kind: ConfigMap
metadata:
  name: myconfigmap
  namespace: default
data:
  mydata: |-
    First line of data
    Second line of data                # with some spaces
```
```
# some random comments
 # some random comments
apiVersion: v1
kind: ConfigMap
metadata:
  name: myconfigmap
  namespace: default # some random comments should be deleted
data:
  mydata: |-
    First line of data # some random comments should not be deleted
    Second line of data    
```
are not handled correctly.

This PR is to make the parsing logic more robust to handle objects parsing correctly.

### CODE REVIEW

File: operator/pkg/object/objects_test.go [hunk 1]
```diff
@@ -712,3 +715,213 @@ func TestK8sObject_ResolveK8sConflict(t *testing.T) {
...
+		name          string
+		manifest      string
+		failOnError   bool
+		expectErr     bool
+		expectCount   int
+		expectedYAMLs []string
+	}{
+		{
+			name: "well formed yaml, no errors",
+			manifest: `
```
Comment (zirain) [inline operator/pkg/object/objects_test.go:730]: can you move these lines to single test file per test case?

Comment (hanxiaop) [reply operator/pkg/object/objects_test.go:730]: Thanks, will do

Comment (hanxiaop) [reply operator/pkg/object/objects_test.go]: Fixed

File: operator/pkg/object/testdata/well-formed-with-space.out.yaml [hunk 2]
```diff
@@ -0,0 +1,9 @@
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: myconfigmap
+  namespace: default
+data:
+  mydata: |-
+    First line of data
+    Second line of data
```
Comment (zirain) [inline operator/pkg/object/testdata/well-formed-with-space.out.yaml]: add new line an the end

Comment (hanxiaop) [reply operator/pkg/object/testdata/well-formed-with-space.out.yaml]: @zirain Fixed, PTAL

**Review by zirain — COMMENTED** [review]
lgtm, defer to @howardjohn

**Review by howardjohn — COMMENTED** [review]
So before:
* We pass raw YAML on, unless we overlay it in which case we re-format it


Now:
* We always reformat

is that accurate?

**Comment by hanxiaop** [comment]
> So before:
> 
> * We pass raw YAML on, unless we overlay it in which case we re-format it
> 
> Now:
> 
> * We always reformat
> 
> is that accurate?

@howardjohn Yes this is correct, overlay will do the same kind of unmarshal and marshal. Is there any concerns?

I think the reformat/overlay parts are all manipulating the yaml field of the object, and the actual apply logic is using the unstructured object directly, so the order won't be a issue.

```
// K8sObject is an in-memory representation of a k8s object, used for moving between different representations
// (Unstructured, JSON, YAML) with cached rendering.
type K8sObject struct {
	object *unstructured.Unstructured

	Group     string
	Kind      string
	Name      string
	Namespace string

	json []byte
	yaml []byte
}
```

I'm not sure if there are historical reasons that we had the cleaning raw YAML logic, but I think it's a bug which may potentially cause parse errors.

**Comment by hanxiaop** [comment]
I actually think the parse/overlay logic is too complicated with extra operations, tpath, etc...

**Comment by hanxiaop** [comment]
Kindly ping @howardjohn

**Review by howardjohn — APPROVED** [review]

**Pull request merged**